### PR TITLE
Use resource files for list of modules and plugins

### PR DESCRIPTION
**Pull request by rjernst**
This adds modules.txt and plugins.txt to the core jar resource files,
which the install plugin command statically loads, in place of the
previously hardcoded lists (which have often gone out of date).

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java [hunk 1]
```diff
@@ -103,37 +102,41 @@
...
+            throw new RuntimeException(e);
+        }
+    }
+
+    /** The official plugins that can be installed simply by name. */
+    static final Set<String> OFFICIAL_PLUGINS;
+    static {
+        try (InputStream stream = InstallPluginCommand.class.getResourceAsStream("/plugins.txt");
+             BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
+            Set<String> plugins = new TreeSet<>(); // use tree set to get sorting for help command
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: Maybe add a test for this?

File: core/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java [hunk 2]
```diff
@@ -103,37 +102,41 @@
...
+            String line = reader.readLine();
+            while (line != null) {
+                plugins.add(line.trim());
+                line = reader.readLine();
+            }
+            plugins.add("x-pack");
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: Too bad. Definitely add a test that this is included?

File: core/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java [hunk 3]
```diff
@@ -103,37 +102,41 @@
...
+            OFFICIAL_PLUGINS = Collections.unmodifiableSet(plugins);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java]: Extraneous blank line here.

**Comment by jasontedor** [comment]
I like it a lot, great improvement. I left a comment about maybe adding some tests for some of the behavior though?

**Comment by rjernst** [comment]
@jasontedor I pushed some more commits adding the tests you asked for.

File: qa/evil-tests/src/test/java/org/elasticsearch/plugins/InstallPluginCommandTests.java [hunk 4]
```diff
@@ -532,6 +534,35 @@ public void testZipRelativeOutsideEntryName() throws Exception {
...
+            // now check each line compares greater than the last, until we reach an empty line
+            String prev = reader.readLine();
+            line = reader.readLine();
+            while (line != null && line.trim().isEmpty() == false) {
+                assertTrue(prev + " < " + line, prev.compareTo(line) < 0);
+                prev = line;
+                line = reader.readLine();
+            }
+        }
+        terminal.getOutput();
```
Comment (jasontedor) [inline qa/evil-tests/src/test/java/org/elasticsearch/plugins/InstallPluginCommandTests.java]: I think this line can be removed?

Comment (rjernst) [inline qa/evil-tests/src/test/java/org/elasticsearch/plugins/InstallPluginCommandTests.java]: Ooops, removed.

**Comment by jasontedor** [comment]
The tests are great. Thank you for adding them. LGTM.

**Pull request merged**